pkg/controller/controller_utils_test.go

```diff
@@ -1121,6 +1124,228 @@ func TestRemoveTaintOffNode(t *testing.T) {
 	}
 }
 
+func TestGetPodsMatchingSelectorUsingPodNamespaceLabelIndex(t *testing.T) {
+	informer := cache.NewSharedIndexInformer(
+		&cache.ListWatch{
+			ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+				return &v1.PodList{}, nil
+			},
+			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+				return watch.NewFake(), nil
+			},
+		},
+		&v1.Pod{},
+		0,
+		cache.Indexers{
+			cache.NamespaceIndex: cache.MetaNamespaceIndexFunc,
+		},
+	)
+
+	if err := AddPodNamespaceLabelIndexer(informer); err != nil {
+		t.Fatalf("failed to add indexer: %v", err)
+	}
+
+	// Add test pods
+	pod1 := &v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: "default",
+			Name:      "pod1",
+			UID:       "uid1",
+			Labels: map[string]string{
+				"app":         "foo",
+				"role":        "backend",
+				"environment": "prod",
+			},
+		},
+	}
+	pod2 := &v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: "default",
+			Name:      "pod2",
+			UID:       "uid2",
+			Labels: map[string]string{
+				"app":         "foo",
+				"role":        "frontend",
+				"environment": "prod",
+			},
+		},
+	}
+	pod3 := &v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: "kube-system",
+			Name:      "pod3",
+			UID:       "uid3",
+			Labels: map[string]string{
+				"app":         "foo",
+				"role":        "backend",
+				"environment": "dev",
+			},
+		},
+	}
+	pod4 := &v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: "default",
+			Name:      "pod4",
+			UID:       "uid4",
+			Labels: map[string]string{
+				"app":                 "foo",
+				"role":                "backend",
+				"special@character":   "value!@#$",
```

Comment:
This is not a valid label per https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/